neuralpredictors/measures/np_functions.py

```diff
@@ -12,10 +19,148 @@ def corr(y1, y2, axis=-1, eps=1e-8, **kwargs):
         y2:      second NumPy array
         axis:    dimension(s) along which the correlation is computed. Any valid NumPy axis spec works here
         eps:     offset to the standard deviation to avoid exploding the correlation due to small division (default 1e-8)
-        **kwargs: passed to final numpy.mean operatoin over standardized y1 * y2
+        **kwargs: passed to final numpy.mean operation over standardized y1 * y2
 
     Returns: correlation array
     """
+
     y1 = (y1 - y1.mean(axis=axis, keepdims=True)) / (y1.std(axis=axis, keepdims=True, ddof=0) + eps)
     y2 = (y2 - y2.mean(axis=axis, keepdims=True)) / (y2.std(axis=axis, keepdims=True, ddof=0) + eps)
     return (y1 * y2).mean(axis=axis, **kwargs)
+
+
+def oracle_corr_corrected(repeated_outputs: ArrayLike) -> np.ndarray:
+    """
+    Compute the corrected oracle correlations per neuron.
+
+    Args:
+        repeated_outputs (array-like): numpy array with shape (images, repeats, responses), or a list of lists of
+            responses (with len: num of repeats) per image.
+
+    Returns:
+        array: Corrected oracle correlations per neuron
+    """
+
+    var_noise, var_mean = [], []
+    for repeat in repeated_outputs:
+        var_noise.append(repeat.var(axis=0))
+        var_mean.append(repeat.mean(axis=0))
+    var_noise = np.mean(np.array(var_noise), axis=0)
+    var_mean = np.var(np.array(var_mean), axis=0)
+    return var_mean / np.sqrt(var_mean * (var_mean + var_noise))
+
+
+def oracle_corr(repeated_outputs: ArrayLike) -> np.ndarray:
+    """
+    Compute the oracle correlations per neuron.
+
+    Args:
+        repeated_outputs (array-like): numpy array with shape (images, repeats, responses), or a list of lists of
+            responses (with len: num of repeats) per image.
+
+    Returns:
+        array: Oracle correlations per neuron
+    """
+
+    if len(repeated_outputs.shape) == 3:
+        _, r, n = repeated_outputs.shape
+        oracles = (repeated_outputs.mean(axis=1, keepdims=True) - repeated_outputs / r) * r / (r - 1)
```

Comment:
Hm... with this, I'd really like for this duplication to be removed. Could you just handle the case of `len(repeated_outputs.shape)==3` as follows:

```
if len(repeated_outputs.shape) == 3:
    # repeated_outputs is not a stack of entries, so turn it into a single element list
    repeated_outputs = [repeated_outputs]

# Now just use the code found under else clause and the rest should work
```

Reply:
I'm not sure if that would work. I think repeated_outputs is a list containing for each image an array of shape (repeats, responses). So I guess it should work with just the code under the else clause: iterate over the list (i.e. for each image). Like this:
`oracles = []
    for outputs in repeated_outputs:
        r, n = outputs.shape
        # compute the mean over repeats, for each neuron
        mu = outputs.mean(axis=0, keepdims=True)
        # compute oracle predictor
        oracle = (mu - outputs / r) * r / (r - 1)

        if np.any(np.isnan(oracle)):
            logger.warning(
                "{}% NaNs when calculating the oracle. NaNs will be set to Zero.".format(
                    np.isnan(oracle).mean() * 100
                )
            )
            oracle[np.isnan(oracle)] = 0

        oracles.append(oracle)
    return corr(np.vstack(repeated_outputs), np.vstack(oracles), axis=0)`

Reply:
Ah, I missed that one, I changed only one occurrence of that structure in l. 40 🤦🏻‍♂️ I'll push the updates to this code portion, too